enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,252 @@
+- The failed installPlan is referenced by a subscription in the namespace.
+- The catalogSource has been updated to include a new new upgrade for the set of arguments.
+
+In practice, the fourth step from the previous workflow would be removed, meaning that the cluster admins would simply need to update the CatalogSource and wait for the cluster to move past the failed install. We do not need to code special handling for partially applied CSVs as OLM will include any applied CSVs in the set of arguments sent to the resolver.
+
+>BEST PRACTICE NOTE: If a bundle is known to fail, it should be skipped in the upgrade graph using the [skips](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skips) or [skipRange](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/#skiprange) feature.
+
+##### Supporting "Fail Forward" Upgrades when an InstallPlan failed
+
+Let's review what steps must be taken to recover from a failed CSV today:
+1. `Operator v1` is being upgraded to `Operator v2`.
+2. The CSV for `Operator v2` enters the FAILED phase.
+3. `Operator v3` is added to the catalog which skips `Operator v2` and replaces `Operator v1`.
+3. The resolver cannot upgrade `Operator v2` while including `Operator v1` in the solution set, upgrade is blocked.
+4. User manually deletes the existing CSVs, a new InstallPlan is generated and approved.
+5. `Operator v3` is installed and the upgrade succeeds.
+
+The enhancement proposes that when the user opts into "fail forward" upgrades that OLM:
+- Does not include CSVs in the REPLACING phase in the set of arguments sent to the Resolver.
+- Allows CSVs to move from the FAILED phase to the REPLACING phase.
+
+In practice, these changes would allow OLM to recover from any number of failed CSVs in an upgrade path. cluster admins would only need to update the CatalogSource with new versions the replace the FAILED CSVs.
+#### Preventing an operator from "Failing Forward" when in the FAILED Phase
+
+This enhancement proposes the introduction of the "fail forward" [Bundle Property](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#properties) to ensure that critical versions of an operator are not skipped by the "fail forward" feature.
+
+To prevent an operator from "failing forward" to the next version, a catalog curator/operator author would need to include the following annotation on the bundle like so:
+
+```yaml=
+image: quay.io/foo/bar:baz
+name: foo-operator.v1.0.1
+package: bar
+properties:
+- type: olm.gvk
+  value:
+    group: foo.io
+    kind: bar
+    version: v1alpha1
+- type: olm.package
+  value:
+    packageName: stable
+    version: 1.0.1
+- type: olm.failForward
+  value:
+    supported: false # This bundle does not support "failing forward" to the next version.
```

Comment:
Two thoughts:
1. Should we have a name that would make the default `false` rather than `true` (e.g. `olm.blockFailForward: true` to opt in). I think that's more natural for people.
2. Instead of an explicit property about whether fail forward should be enabled or not, what if we named this something like `olm.performsCriticalMigration` or `olm.hasCriticalSection` or something along those lines?

(2) Is a little more specific and could leave us wiggle room in the future to make the on-cluster upgrade strategy logic consider more than one property. We could also implement some interesting validation logic for this property (e.g. only `X.Y.0` releases are allowed to set `olm.performsCriticalMigration` to `true`.

Reply:
The downside of my suggestion (2) is that there would no longer be a clear tie-in with the on-cluster behavior, so maybe that's a reason to leave it as is?

Reply:
No longer relevant based on [this comment](https://github.com/operator-framework/enhancements/pull/110/files#r837906801)